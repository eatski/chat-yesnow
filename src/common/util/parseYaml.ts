import { parse } from "yaml";
import { z } from "zod";

const answer = z.enum(["はい", "いいえ", "わからない"]).transform((text) => {
	switch (text) {
		case "はい":
			return "True";
		case "いいえ":
			return "False";
		case "わからない":
			return "Unknown";
	}
});

const questionExample = z.object({
	question: z.string(),
	answer,
	supplement: z.string().optional(),
	customMessage: z.string().optional(),
});

export const storyInitYaml = z.object({
	title: z.string(),
	quiz: z.string(),
	truth: z.string(),
	simpleTruth: z.string(),
	questionExamples: z.array(questionExample),
});

type ParseResult =
	| {
			error: null;
			data: z.infer<typeof storyInitYaml>;
	  }
	| {
			error: string;
	  };

export const parseYaml = (yaml: string): ParseResult => {
	try {
		parse(yaml);
		const parsed = parse(yaml);
		const typed = storyInitYaml.safeParse(parsed);
		if (!typed.success)
			return {
				error: `値に不備もしくは不足があります: ${typed.error.errors
					.flatMap((e) => e.path)
					.join(",")}`,
			};
		return {
			error: null,
			data: typed.data,
		};
	} catch (e) {
		return {
			error: "YAMLの形式が不正です。",
		};
	}
};
