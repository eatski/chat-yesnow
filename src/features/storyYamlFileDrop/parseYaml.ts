import { parse } from "yaml";
import { storyInitYaml } from "./type";
import { z } from "zod";

type ParseResult =
	| {
			error: null;
			data: z.infer<typeof storyInitYaml>;
	  }
	| {
			error: string;
	  };

export const parseYaml = (yaml: string): ParseResult => {
	try {
		parse(yaml);
		const parsed = parse(yaml);
		const typed = storyInitYaml.safeParse(parsed);
		if (!typed.success)
			return {
				error: `値に不備もしくは不足があります: ${typed.error.errors
					.flatMap((e) => e.path)
					.join(",")}`,
			};
		return {
			error: null,
			data: typed.data,
		};
	} catch (e) {
		return {
			error: "YAMLの形式が不正です。",
		};
	}
};
