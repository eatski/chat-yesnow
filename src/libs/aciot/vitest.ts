import { AxiosInstance } from "axios";
import { AciotCore, CacheMode } from "./core";
import { afterAll, beforeAll } from "vitest";
import path from "path";
import filenamify from 'filenamify';

export const applyTestHooks = (
	axios: AxiosInstance,
	config: {
		mode: CacheMode;
	},
) => {
	let aciot: AciotCore | null;
	beforeAll(({ name, file }) => {
		if (!file?.name) {
			throw new Error("suite.file.name is not defined");
		}
		const parsedPath = path.parse(file.name);
		const filename = filenamify(name, { replacement: "_" });
		aciot = new AciotCore({
			cacheBasePath: path.resolve(
				process.cwd(),
				parsedPath.dir,
				"__data__",
				parsedPath.name + parsedPath.ext,
				filename
			),
		});
		aciot.applyInterceptors(axios, config.mode);
	});
	afterAll(async ({ name, tasks }) => {
		if (tasks.every((t) => t.result?.state === "pass")) {
			console.info(name, "Clearing unused cache...");
			await aciot?.clearUnusedCache();
		}
	});
};
