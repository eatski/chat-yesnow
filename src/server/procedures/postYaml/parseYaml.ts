import { parse } from 'yaml'
import { storyInput } from "./type";

export const parseYaml = (yaml: string) => {
    const parsed = parse(yaml);
    const typed = storyInput.safeParse(parsed);
    return typed;
}