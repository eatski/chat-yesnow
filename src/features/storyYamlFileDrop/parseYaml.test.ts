import { readFileSync } from "fs"
import { resolve } from "path"
import {describe,test,expect} from "vitest"
import { parseYaml } from "./parseYaml" 

const sampleYaml = readFileSync(resolve(process.cwd(),"samples","sample1.yaml"), "utf-8");
const errorYaml = readFileSync(resolve(process.cwd(),"samples","error1.yaml"), "utf-8");

describe('parseYaml', () => {
    test("should parse input yaml", ()=> {
        expect(parseYaml(sampleYaml)).toMatchSnapshot()
    })
    test("should return error when input is invalid", ()=> {
        expect(parseYaml(errorYaml)).toMatchSnapshot()
    })
})