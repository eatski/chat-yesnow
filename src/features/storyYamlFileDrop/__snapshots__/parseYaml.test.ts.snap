// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseYaml > should parse input yaml 1`] = `
{
  "data": {
    "questionExamples": [
      {
        "answer": "False",
        "question": "yesonorの裏側で開発者が頑張って回答を考えている？",
        "supplement": "頑張っているのはAIであり、開発者ではありません。",
      },
      {
        "answer": "True",
        "question": "yesonorでは回答者の質問に対してAIが回答を考えている？",
        "supplement": "yesonorではGPT-4によって質問に対する回答を生成しています。",
      },
      {
        "answer": "Unknown",
        "question": "yesonorの開発者はこのサイトを作るのに苦労した？",
        "supplement": "開発者の苦労については言及されておりません。",
      },
    ],
    "quiz": "水平思考クイズとは通常、真相を知る「出題者」と質問をして謎を解き明かす「回答者」とで遊ぶゲームです。
しかし、Yesonorでは水平思考クイズを回答者のみで遊べます。なぜでしょう？
",
    "simpleTruth": "YesonorではAIが質問に対する回答を行う。",
    "title": "Yesonorの謎",
    "truth": "YesonorではAI（GPT-4）が質問に対する回答を行うため出題者が不要です。",
  },
  "error": null,
}
`;

exports[`parseYaml > should return error when input is invalid 1`] = `
{
  "error": "値に不備もしくは不足があります: truth,simpleTruth",
}
`;
